src/Features/Core/Portable/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -79,57 +87,58 @@ public DesignerAttributeDiscoveryService()
             if (!project.SupportsCompilation)
                 return;
 
-            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var designerCategoryType = compilation.DesignerCategoryAttributeType();
-            if (designerCategoryType == null)
-                return;
+            // Defer expensive work until it's actually needed.
+            var lazyProjectVersion = AsyncLazy.Create(project.GetSemanticVersionAsync, cacheResult: true);
+            var lazyHasDesignerCategoryType = s_metadataReferencesToDesignerAttributeInfo.GetValue(
+                project.MetadataReferences,
+                _ => AsyncLazy.Create(
+                    async cancellationToken =>
+                    {
+                        var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                        return compilation.DesignerCategoryAttributeType() != null;
+                    }, cacheResult: true));
```

Comment:
figuring out if a set of references contains the DesignCategoryAttribute can be computed/cached for that set of references.  As those rarely change for a project, it's much more efficient than having to continually use the project and create the compilation for it.